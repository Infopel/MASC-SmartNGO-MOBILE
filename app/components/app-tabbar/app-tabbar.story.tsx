import * as React from "react"
import { storiesOf } from "@storybook/react-native"
import { StoryScreen, Story, UseCase } from "../../../storybook/views"
import { AppTabbar } from "./app-tabbar"

storiesOf("AppTabbar", module)
  .addDecorator((fn) => <StoryScreen>{fn()}</StoryScreen>)
  .add("Style Presets", () => (
    <Story>
      <UseCase text="Primary" usage="The primary.">
        <AppTabbar m='2' />
      </UseCase>
    </Story>
  ))
