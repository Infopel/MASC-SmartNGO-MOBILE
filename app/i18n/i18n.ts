import * as Localization from "expo-localization"
//@ts-ignore
import {I18n} from "i18n-js/dist/require"
import en from "./en"
import pt from "./pt"
export const i18n =  new I18n({en,pt})
i18n.enableFallback = true
i18n.defaultLocale = "en"
// i18n.missingBehavior = "error";

i18n.locale = Localization.locale || "en"

/**
 * Builds up valid keypaths for translations.
 * Update to your default locale of choice if not English.
 */
type DefaultLocale = typeof en
export type TxKeyPath = RecursiveKeyOf<DefaultLocale>

type RecursiveKeyOf<TObj extends Record<string, any>> = {
  [TKey in keyof TObj & string]: TObj[TKey] extends Record<string, any>
  //@ts-ignore  
  ? `${TKey}` | `${TKey}.${RecursiveKeyOf<TObj[TKey]>}`
    : `${TKey}`
}[keyof TObj & string]
