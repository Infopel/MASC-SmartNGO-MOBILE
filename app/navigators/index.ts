export * from "./root-navigator"
export * from "./navigation-utilities"
// export other navigators from here
export * from "./navigators.d"
export * from "./home-bottom-tab/home-bottom-tab-navigator"
export * from "./auth/auth-navigator"
export * from "./primary/primary-navigator"
export * from "./civic-incubator-details-tab/civic-incubator-details-tab-navigator"
export * from "./savings-group-details-tab/savings-group-details-tab-navigator"
