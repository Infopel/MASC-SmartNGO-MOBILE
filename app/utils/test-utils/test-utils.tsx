import { render } from "@testing-library/react-native"
import React from "react"
import { NativeBaseProvider as ThemeProvider } from "theme"
import { AllTheProvidersProps, RenderOptions } from "./render"

function AllTheProviders({ children }: AllTheProvidersProps) {
  return (
    <ThemeProvider>{children}</ThemeProvider>
  )
}

function customRender<T extends any>(ui: React.ReactElement<T>, options?: RenderOptions) {
  return render(ui, { wrapper: AllTheProviders, ...options })
}

// re-export everything
export * from "@testing-library/react-native"
// override render method
export { customRender as render }

