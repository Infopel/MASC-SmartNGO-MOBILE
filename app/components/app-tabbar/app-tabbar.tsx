import { useToolbarColor } from "components/hooks"
import { Button as NBButton, HStack, useColorModeValue } from "native-base"
import { TabBarProps } from "navigators/navigators"
import * as React from "react"
import { TabBarButton } from "./tab-bar-button"

export interface AppTabbarProps extends TabBarProps {

}

/**
 * Describe your component here
 */
export function AppTabbar({ state, jumpTo,navigation: { navigate } }: AppTabbarProps) {

  const routes = state.routes
const {bg,border} = useToolbarColor()
  console.log({jumpTo})

  return (
    <HStack
      m='0'
      bg={bg}
      borderLeftColor={border}
      borderBottomColor={border}
      borderRightColor={border}
      borderTopColor={bg}
      borderBottomWidth='1'
      shadow='3'
      alignItems={'center'}
      justifyContent='space-evenly'
      w='full'
      py='2'
    >
      {routes.map((item, index) => <TabBarButton tx={item.name} selected={index === state.index} onPress={jumpTo.bind(this, item.key, undefined)} />)}
    </HStack>
  )
}

