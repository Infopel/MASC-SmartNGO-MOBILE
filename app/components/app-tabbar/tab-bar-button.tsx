import { Button } from "components/button/button";
import { useIsPressed } from "components/hooks";
import { TxKeyPath } from "i18n/i18n";
import { IBoxProps } from "native-base";
import React from "react";

interface TabBarButtonProps extends IBoxProps {
    onPress: () => void;
    tx: TxKeyPath
    selected: boolean

}
export function TabBarButton({ selected, tx, ...rest }: TabBarButtonProps) {
    const { pressableProps, isPressed } = useIsPressed()


    return <Button tx={`${tx}.title`} _text={{
        fontWeight: 'medium'
    }}  {...rest} p='2' borderRadius='full' variant={selected ? 'solid' : 'ghost'} isPressed={isPressed} {...pressableProps} accessibilityState={{ selected: selected }} />
}