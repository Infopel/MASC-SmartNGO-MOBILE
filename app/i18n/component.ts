import {TranslateOptions} from 'i18n-js/typings';
import logger from 'utils/logger';
import {TxKeyPath} from './i18n';
import {translate} from './translate';

export interface ITranslatable {
  /**
   * Optional options to pass to i18n. Useful for interpolation
   * as well as explicitly setting locale or translation fallbacks.
   */
  txOptions?: TranslateOptions;
}
export interface ITranslatableText extends ITranslatable {
  /**
   * Text which is looked up via i18n.
   */
  tx?: TxKeyPath;

  /**
   * Optional options to pass to i18n. Useful for interpolation
   * as well as explicitly setting locale or translation fallbacks.
   */
  txOptions?: TranslateOptions;

  /**
   * The text to display if not using `tx` or nested components.
   */

  children?: React.ReactNode;
}

export function translateComponent({
  tx,
  children,
  txOptions,
}: ITranslatableText): React.ReactNode | undefined {
  const i18nText = tx && translate(tx, txOptions);

  if (typeof i18nText !== 'string' && i18nText !== undefined) {
    console.warn('Invalid translated text', {i18nText});
    return children;
  }
  return children || i18nText;
}
