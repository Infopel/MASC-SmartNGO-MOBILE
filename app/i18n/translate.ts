import {TranslateOptions} from 'i18n-js/typings';
import {i18n} from './i18n';
import {TxKeyPath} from './i18n';

/**
 * Translates text.
 *
 * @param key The i18n key.
 */
export function translate(key: TxKeyPath, options?: TranslateOptions) {
  if (!key) throw new Error('Key cannot be nil or empty');
  const trans = i18n.t(key, options);

  if (typeof trans !== 'string')
  throw new Error("The inserted translation key should not map to and object");
  
  return trans;
}
