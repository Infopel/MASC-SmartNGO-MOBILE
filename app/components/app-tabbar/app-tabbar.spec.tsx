import { TxKeyPath } from "i18n/i18n"
import { translate  as tx } from "i18n/translate"
import { TabBarProps } from "navigators/navigators"
import * as React from "react"
import { fireEvent, render } from "utils/test-utils/test-utils"
import { AppTabbar } from "./app-tabbar"

let nav = {
  jumpTo: key => nav = { ...nav, state: { ...nav.state, index: nav.state.routeNames.indexOf((s) => key === s) } },
  state: {
    index: 0,
    routeNames: [
      "civic-incubator-details-events-tab", "civic-incubator-details-participants-tab",
    ],
  }
} as TabBarProps


describe("<AppTabbar>", () => {
  it("render", async function () {
    const component = <AppTabbar {...nav}></AppTabbar>

    expect(render(component).toJSON()).toMatchSnapshot()
  })

  it("should render toolbar buttons", async function () {

    const { getByText, getAllByRole } = render(<AppTabbar {...nav}></AppTabbar>)
    const routes = nav.state.routeNames
    expect(getAllByRole('button').length).toBe(routes.length)

    for (const item of routes) {
      expect(getByText(translate(item))).toBeTruthy()
    }
  })

  it("should have 1st button selected", async function () {

    const routes = nav.state.routeNames
    const { getByText, queryByAccessibilityState } = render(<AppTabbar {...nav}></AppTabbar>)
    expect(queryByAccessibilityState({ selected: true })).toHaveTextContent(translate(routes[0]))
  })

  it("should select 2nd button", async function () {

    const routes = nav.state.routeNames
    const index = 1
    const { getByText, queryByAccessibilityState } = render(<AppTabbar {...nav}></AppTabbar>)
    expect(queryByAccessibilityState({ selected: true })).toHaveTextContent(translate(routes[0]))

    fireEvent.press(getByText(translate(routes[index])))
    expect(queryByAccessibilityState({ selected: true })).toHaveTextContent(translate(routes[index]))
  })
})

function translate(key: TxKeyPath){
  return  tx(key+".title")
}

