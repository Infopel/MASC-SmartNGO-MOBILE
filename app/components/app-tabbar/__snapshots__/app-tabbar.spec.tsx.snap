// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppTabbar> render 1`] = `
<View
  dataSet={Object {}}
  style={
    Object {
      "marginBottom": 16,
      "marginLeft": 16,
      "marginRight": 16,
      "marginTop": 16,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Array [
        Object {},
        Object {
          "dataSet": Object {},
        },
      ]
    }
    data={
      Array [
        "civic-incubator-details-events-tab",
        "civic-incubator-details-participants-tab",
      ]
    }
    dataSet={Object {}}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={Object {}}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          _icon={
            Object {
              "color": "text.50",
              "size": "sm",
            }
          }
          _spinner={
            Object {
              "color": "text.50",
              "focusable": false,
              "size": "sm",
            }
          }
          _stack={
            Object {
              "alignItems": "center",
              "space": "1.5",
            }
          }
          _text={
            Object {
              "color": "text.50",
              "fontSize": "sm",
            }
          }
          accessibilityRole="button"
          accessibilityState={
            Object {
              "selected": true,
            }
          }
          accessible={true}
          collapsable={false}
          colorScheme="primary"
          dataSet={Object {}}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onHoverIn={[Function]}
          onHoverOut={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d37b16",
              "borderRadius": 4,
              "flexDirection": "row",
              "justifyContent": "center",
              "paddingBottom": 10,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 10,
            }
          }
        >
          <View
            dataSet={Object {}}
            style={Object {}}
          >
            <Text
              dataSet={Object {}}
              style={
                Object {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 24,
                  "fontStyle": "normal",
                  "fontWeight": "700",
                  "letterSpacing": 0,
                  "lineHeight": 30,
                  "textDecorationLine": undefined,
                }
              }
            >
              Lectures
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          _icon={
            Object {
              "color": "primary.900",
              "size": "sm",
            }
          }
          _spinner={
            Object {
              "color": "primary.900",
              "focusable": false,
              "size": "sm",
            }
          }
          _stack={
            Object {
              "alignItems": "center",
              "space": "1.5",
            }
          }
          _text={
            Object {
              "color": "primary.900",
              "fontSize": "sm",
            }
          }
          accessibilityRole="button"
          accessibilityState={
            Object {
              "selected": false,
            }
          }
          accessible={true}
          collapsable={false}
          colorScheme="primary"
          dataSet={Object {}}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onHoverIn={[Function]}
          onHoverOut={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffcc91",
              "borderRadius": 4,
              "flexDirection": "row",
              "justifyContent": "center",
              "paddingBottom": 10,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 10,
            }
          }
        >
          <View
            dataSet={Object {}}
            style={Object {}}
          >
            <Text
              dataSet={Object {}}
              style={
                Object {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 24,
                  "fontStyle": "normal",
                  "fontWeight": "700",
                  "letterSpacing": 0,
                  "lineHeight": 30,
                  "textDecorationLine": undefined,
                }
              }
            >
              Participants
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`<AppTabbar> should select 2nd button 1`] = `
<View
  dataSet={Object {}}
  style={
    Object {
      "marginBottom": 16,
      "marginLeft": 16,
      "marginRight": 16,
      "marginTop": 16,
    }
  }
>
  <View
    _icon={
      Object {
        "color": "text.50",
        "size": "sm",
      }
    }
    _spinner={
      Object {
        "color": "text.50",
        "focusable": false,
        "size": "sm",
      }
    }
    _stack={
      Object {
        "alignItems": "center",
        "space": "1.5",
      }
    }
    _text={
      Object {
        "color": "text.50",
        "fontSize": "sm",
      }
    }
    accessible={true}
    collapsable={false}
    colorScheme="primary"
    dataSet={Object {}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onHoverIn={[Function]}
    onHoverOut={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#d37b16",
        "borderRadius": 4,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingTop": 10,
      }
    }
  >
    <Text
      dataSet={Object {}}
      style={
        Object {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 21,
          "textDecorationLine": undefined,
        }
      }
    >
      Lectures
    </Text>
  </View>
  <View
    _icon={
      Object {
        "color": "text.50",
        "size": "sm",
      }
    }
    _spinner={
      Object {
        "color": "text.50",
        "focusable": false,
        "size": "sm",
      }
    }
    _stack={
      Object {
        "alignItems": "center",
        "space": "1.5",
      }
    }
    _text={
      Object {
        "color": "text.50",
        "fontSize": "sm",
      }
    }
    accessible={true}
    collapsable={false}
    colorScheme="primary"
    dataSet={Object {}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onHoverIn={[Function]}
    onHoverOut={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#d37b16",
        "borderRadius": 4,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingTop": 10,
      }
    }
  >
    <Text
      dataSet={Object {}}
      style={
        Object {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 21,
          "textDecorationLine": undefined,
        }
      }
    >
      Participants
    </Text>
  </View>
</View>
`;
